Don't commit entries from previous terms by counting replicas + noop entry upon leader election

MathieuBordere opened the pull request:
An entry from a previous term will no longer be committed by counting votes (which violates the raft spec), instead upon leader election a leader applies a noop barrier entry to its log. When committing the noop barrier entry, all previous entries in the log will be committed.

The noop entry upon leader election is needed because raft, on startup doesn't know if its latest configuration loaded from disk is committed or not, therefore it must set `configuration_uncommitted_index` (to be added in following PR) and will reject any configuration changes requested by the client. To clear `configuration_uncommitted_index` the entry at that index must be committed. It's also needed when a follower becomes leader while still having a non-0 `configuration_uncommitted_index`.

This also includes a fix to update `last_applied` before firing an applied request's callback. The callback could contain logic dependent on `last_applied` which should have logically increased after applying the entry, but before firing the callback. This fixes an issue in dqlite where a barrier could fire another barrier in its callback.

fixes https://github.com/canonical/raft/issues/220.

! This will break the dqlite tests, but not dqlite's functionality, I've updated raft's version so dqlite's configure step can test against it.

MathieuBordere commented on src/replication.c:
```diff
@@ -1649,19 +1669,21 @@ void replicationQuorum(struct raft *r, const raft_index index)
+     * of matchIndex[i] >= N, and log[N].term == currentTerm:
+     * set commitIndex = N (5.3, 5.4).
+     * */
+    for (n = index; n > r->commit_index; n -= 1) {
```
This implementation might be a bit naive, if n is large and commit index is 0 e.g. and a majority of the followers their match index hasn't been updated or is low, this loop can run for a long while, I'll first fix this.

MathieuBordere commented:
Converted to draft to optimize the vote counting a bit first.

MathieuBordere commented:
Added an optimization, need to test it a bit more to feel comfortable to release it for review. In short, raft will only try to commit log entry indices that are smaller than or equal to the largest match_index of another voter in the cluster (except in some corner cases) 

This fixes a case where the loop that probes for new commit indices would run needlessly for e.g. a 1000 times when the leaders match index is for example 5000 and the next voter's match index is only 4000. In this case it only makes sense to start looking for a new commit index from 4000 and below.

freeekanayaka commented on src/convert.c:
```diff
@@ -180,6 +180,12 @@ int convertToCandidate(struct raft *r, bool disrupt_leader)
     return 0;
 }
 
+void leader_noop_barrier_cb(struct raft_barrier *req, int status)
```
Nit: I'd name this `convertInitialBarrierCb` or something like that, for consistency with the naming convention (camelCase, and private function names start with the module name).

MathieuBordere replied:
hehe, I just keep sticking to the underscores, will try to remember next time :-)

freeekanayaka reviewed (approved):
Looks good to me at first sight. Just a new minor nits.

freeekanayaka commented on src/replication.c:
```diff
@@ -1630,38 +1634,93 @@ int replicationApply(struct raft *r)
-    assert(r->state == RAFT_LEADER);
+    unsigned i;
+    unsigned server_index;
+    raft_index idx;
```
Nit: `idx` -> `index` for consistency with the convention (full words).

MathieuBordere commented on src/replication.c:
```diff
@@ -1630,38 +1634,93 @@ int replicationApply(struct raft *r)
-    if (index <= r->commit_index) {
-        return;
+    /* In a single voter cluster where this node is part of the configuration,
+     * is a voter all stored entries can be committed. */
```
Thanks, will investigate if it can be dropped.

freeekanayaka replied:
> Thanks, will investigate if it can be dropped.

Not sure it's clear, I mean that the ***words*** `is a voter` can be dropped from the comment (probably they are a leftover? the full comment does quite sound grammatically correct to me).

MathieuBordere replied:
Ah okay, yes the sentence indeed looks a bit strange.

MathieuBordere replied:
tests also still pass when I remove `r->configuration.servers[server_index].role == RAFT_VOTER` from the check, I will investigate anyway to make sure my understanding is correct.

freeekanayaka replied:
> tests also still pass when I remove `r->configuration.servers[server_index].role == RAFT_VOTER` from the check, I will investigate anyway to make sure my understanding is correct.

I was a bit suspicious about that condition too, but didn't mention it actually. I would not expect that cases where there is a single server and that server is not a voter to be possible, because otherwise there would be no way to make progress and commit further entries. So that condition could be turned into an assertion. However, I'm just hand-waving, not sure.

MathieuBordere replied:
The check actually says that there is 1 voter in the cluster, not that there is a single server. Checking for `r->configuration.servers[server_index].role == RAFT_VOTER` then means that this server was not demoted to standby or spare in the latest configuration.

freeekanayaka replied:
Ah yeah sorry, I misread the `configurationVoterCount` call, that's the number of voters indeed, not servers. Makes sense then.

MathieuBordere replied:
Ok, I double-checked and I think the condition is fine. The test still succeed after removing the `RAFT_VOTER` condition because the test is less strict then and the `start_index` is chosen to be `last_stored` on the current leader, which will be the highest possible index that can be committed anyway.

MathieuBordere commented on src/replication.c:
```diff
@@ -1630,38 +1634,93 @@ int replicationApply(struct raft *r)
+     * of matchIndex[i] >= N, and log[N].term == currentTerm:
+     * set commitIndex = N (5.3, 5.4).
+     * */
+    start_index = replicationCalculateCommitIndex(r);
```
Yeah, I think it's worth it to optimize this, let me think about it for a bit.

MathieuBordere replied:
I actually think we can keep the current (what's on master) implementation of `replicationQuorum`, while adding the check to not commit entries if the term is older than the current term, on the condition that we call it here https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L754 with `last_index` instead of `r->last_stored`. `last_index` is equal to the `match_index` of the server from which we have just received an AppendEntriesRPC. (due to https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L722)

This works because a new `commit_index` is always equal to some server's `match_index` and all `match_index` updates trickle down to the `replicationQuorum` call. Unless this call https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L746 fails, but then we convert to `RAFT_UNAVAILABLE` anyway.

It's a lot simpler solution than the current one proposed in this PR (and still passes the new tests). Let me know what you think.

MathieuBordere replied:
> That reasoning seems sound to me, and I like the simplification. I guess that here
> 
> https://github.com/canonical/raft/blob/0ddb3296fd4ec7d2e8d62b40aa12a0cb3a49e37e/src/replication.c#L520-L521
> 
> we'd still use r->last_stored, right?

yes

MathieuBordere commented on configure.ac:
```diff
@@ -1,5 +1,5 @@
 AC_PREREQ(2.60)
-AC_INIT([raft], [0.17.0])
+AC_INIT([raft], [0.17.1])
```
Yeah, I want to be able to check this version in `dqlite`'s configure step because this version will break dqlite's tests.

MathieuBordere commented on src/replication.c:
```diff
@@ -1354,6 +1354,9 @@ static int applyCommand(struct raft *r,
         return rv;
     }
+
+    r->last_applied = index;
```
Makes sense to document it, yes.

MathieuBordere commented:
Implemented the simplification described [here](https://github.com/canonical/raft/pull/336#discussion_r1049351311). Removed the test that checks that we don't commit entries from previous terms by counting votes, as it was really testing nothing anymore after this change (it was hacky anyway) and from the code it's clear we can't commit entries from an older term by counting voters.

freeekanayaka commented:
It's surely simpler :)
